# SQL Analysis

In [7]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [8]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

### Data Exploration

1. This query counts the number of tracks on the Ultimate Indie playlist with a popularity greater than or equal to 50. This can be an important KPI to compare to other indie playlists to determine if it is a friendlier space for newer artists, or an indicator that the playlist isn't performing as well as others in its genre.

In [10]:
%%sql
SELECT
	COUNT(track_id) AS TrackCount
FROM ultimate_indie ui
WHERE popularity >= 50;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
1 rows affected.


[(72,)]

As we can see here, the Ultimate Indie playlist has a healthy amount of songs with popularity indexes above 50. This means that while most of the songs are generating a good amount of streams on a consistent basis, the playlist also cultivates a space for newer artists or tracks to generate streams in accordance with the more mainstream artists.

# ----------------------------------------------------------------------------------------

2. This query looks at the average duration of songs on the Ultimate Indie playlist (in minutes) by artist. The query also looks at the average popularity of an artist's songs. These are important metrics to consider whether shorter songs have higher popularities on average.

In [18]:
%%sql
SELECT
	artist,
	FORMAT((AVG(duration) / 1000) / 60 , 2) AS AvgArtistDuration,
	FORMAT(AVG(popularity), 0)
FROM ultimate_indie ui
GROUP BY artist
ORDER BY FORMAT((AVG(duration) / 1000) / 60 , 2);

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
73 rows affected.


[('Porches', '2.06', '42'),
 ('Dijon', '2.12', '58'),
 ('Ritt Momney', '2.44', '50'),
 ('Dehd', '2.51', '60'),
 ('binki', '2.59', '54'),
 ('PawPaw Rod', '2.61', '59'),
 ('Toro y Moi', '2.68', '61'),
 ('Surf Curse', '2.71', '61'),
 ('JAWNY', '2.80', '51'),
 ('Still Woozy', '2.82', '64'),
 ('Sure Sure', '2.83', '54'),
 ('Jordana', '2.88', '63'),
 ('NEIL FRANCES', '2.90', '60'),
 ('Inner Wave', '2.91', '58'),
 ('Men I Trust', '2.94', '66'),
 ('Jungle', '2.94', '57'),
 ('Sipper', '2.97', '49'),
 ('Oso Oso', '3.07', '53'),
 ('Phoebe Bridgers', '3.08', '6'),
 ('Vansire', '3.10', '56'),
 ('Yves Tumor', '3.12', '57'),
 ('The Avalanches', '3.13', '65'),
 ('Crumb', '3.21', '0'),
 ('Kate Bollinger', '3.22', '52'),
 ('Spoon', '3.22', '61'),
 ('Japanese Breakfast', '3.26', '5'),
 ('Rex Orange County', '3.27', '76'),
 ('Andy Shauf', '3.27', '42'),
 ('Omar Apollo', '3.29', '0'),
 ('Goth Babe', '3.31', '59'),
 ('alt-J', '3.32', '58'),
 ('Ginger Root', '3.36', '55'),
 ('Wet Leg', '3.37', '61'),
 ('KUNZITE', '3.41', '48'),
 ('Bonobo', '3.50', '62'),
 ('Dayglow', '3.51', '52'),
 ('Yot Club', '3.53', '54'),
 ('TV Girl', '3.57', '84'),
 ('Unknown Mortal Orchestra', '3.62', '66'),
 ('Metronomy', '3.62', '61'),
 ('Peach Pit', '3.70', '63'),
 ('Foals', '3.74', '70'),
 ('Sports', '3.76', '49'),
 ("Barry Can't Swim", '3.78', '55'),
 ('Clairo', '3.82', '77'),
 ('Tame Impala', '3.83', '65'),
 ('Yumi Zouma', '3.88', '48'),
 ('Bakar', '3.90', '62'),
 ('Arlo Parks', '3.92', '56'),
 ('Magdalena Bay', '3.93', '4'),
 ('Helado Negro', '3.97', '42'),
 ('Hether', '4.07', '49'),
 ('Hovvdy', '4.19', '20'),
 ('Sufjan Stevens', '4.20', '55'),
 ('Big Thief', '4.21', '55'),
 ('Maggie Rogers', '4.21', '0'),
 ('Hippo Campus', '4.23', '62'),
 ("Her's", '4.25', '80'),
 ('Beach House', '4.35', '35'),
 ('DOPE LEMON', '4.37', '67'),
 ('Car Seat Headrest', '4.57', '49'),
 ('The Smile', '4.60', '63'),
 ('Angel Olsen', '4.61', '61'),
 ('Broken Social Scene', '4.72', '44'),
 ('Magic City Hippies', '4.74', '49'),
 ('Khruangbin', '4.87', '72'),
 ('Fontaines D.C.', '5.10', '41'),
 ('DARKSIDE', '5.24', '49'),
 ('Spellling', '5.66', '0'),
 ('Burial', '5.70', '63'),
 ('Black Country, New Road', '6.06', '63'),
 ('King Gizzard & The Lizard Wizard', '6.08', '61'),
 ('Kurt Vile', '7.30', '60')]

Based on this query, it might be inferred that shorter songs in the range between 2 to 4 minutes tend to have slightly higher popularity indexes on this playlist, however it is important to consider that the average function is skewing data grouped by artists rather than tracks. 

# ----------------------------------------------------------------------------------------

3. This query creates a view table for tracks on the Ultimate Indie and Bedroom Pop playlists that contain explicit lyrics. The query below it proceeds to discover explicit tracks with popularities below 50.

In [12]:
%%sql
CREATE OR REPLACE VIEW explicit_track AS
	SELECT * FROM ultimate_indie
	WHERE explicit = 1
	UNION
	SELECT * FROM bedroom_pop
	WHERE explicit = 1;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
0 rows affected.


[]

In [14]:
%%sql
SELECT 
	artist,
	name,
	explicit,
	popularity
FROM explicit_track
WHERE popularity < 50
ORDER BY popularity DESC;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
9 rows affected.


[('Sipper', 'Fuck June', '1', 49),
 ('KUNZITE', 'HALOHEAD', '1', 48),
 ('Wet Leg', 'Ur Mum', '1', 47),
 ('Fontaines D.C.', 'I Love You', '1', 41),
 ('Jordan Ward', 'Lil Baby Crush', '1', 6),
 ('Still Woozy', 'Lucy', '1', 0),
 ('Still Woozy', 'Goodie Bag', '1', 0),
 ('Cuco', 'Lover Is a Day', '1', 0),
 ('Ralph Castelli', 'Bends', '1', 0)]

Based on the results, there are 9 tracks with popularities below 50. 5 of which have popularities below 10. It can be inferred that tracks with explicit lyrics may not perform as well in this playlist, or that it is taking them a longer amount of time to generate streams because people skip the tracks before they can be fully played.

# ----------------------------------------------------------------------------------------

4. This query finds the number of tracks on the Bedroom Pop playlist from each artist. This can be a good way to determine which artists are being featured the most on the playlist.

In [15]:
%%sql
SELECT
	artist,
	COUNT(track_id) AS TrackCount
FROM bedroom_pop bp
GROUP BY artist
ORDER BY COUNT(track_id) DESC;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
67 rows affected.


[('Men I Trust', 7),
 ('Still Woozy', 5),
 ('CASTLEBEAT', 4),
 ('Ralph Castelli', 4),
 ('Cuco', 3),
 ('Mild High Club', 2),
 ('SALES', 2),
 ('Los Retros', 2),
 ('Will Paquin', 2),
 ('RICEWINE', 2),
 ('The Marías', 2),
 ('Sunset Rollercoaster', 2),
 ('Roy Blair', 2),
 ('Dayglow', 2),
 ('TEMPOREX', 2),
 ('Current Joys', 2),
 ('Hether', 2),
 ('Monsune', 2),
 ('Gus Dapperton', 2),
 ('Yot Club', 2),
 ('Teenage Priest', 1),
 ('Baird', 1),
 ('The Simps', 1),
 ('Frankie Cosmos', 1),
 ('fanclubwallet', 1),
 ('dvr', 1),
 ('No Vacation', 1),
 ('KALI', 1),
 ('Clairo', 1),
 ('maxime.', 1),
 ('Kate Bollinger', 1),
 ('Fleece', 1),
 ('Metronomy', 1),
 ('Steve Lacy', 1),
 ('Zack Villere', 1),
 ('DLG.', 1),
 ('Dwyer', 1),
 ('Jordan Ward', 1),
 ('808vic', 1),
 ('Verzache', 1),
 ('Yung Heazy', 1),
 ('Lowertown', 1),
 ('Mac DeMarco', 1),
 ('Jakob', 1),
 ('chloe moriondo', 1),
 ('grentperez', 1),
 ('Good Morning', 1),
 ('Wabie', 1),
 ('George Clanton', 1),
 ('Ginger Root', 1),
 ('HOMESHAKE', 1),
 ('boy pablo', 1),
 ('Hala', 1),
 ('Mellow Fellow', 1),
 ('Vansire', 1),
 ('Strawberry Guy', 1),
 ('Orion Sun', 1),
 ('banzai florist', 1),
 ('MICHELLE', 1),
 ('the booyah! kids', 1),
 ('MyKey', 1),
 ('Far Caspian', 1),
 ('Jack Stauber', 1),
 ('Dev Lemons', 1),
 ('pizzagirl', 1),
 ('Bearcubs', 1),
 ('Sports', 1)]

According to the results, Men I Trust and Still Woozy are two artists with a significant amount of tracks on the Bedroom Pop playlist. Advertising this playlist to their fans would be a great way to bring more monthly listeners to the playlist.

# ----------------------------------------------------------------------------------------

5. Finally, this query looks at the total duration of the Surf Rock Sunshine playlist, in hours. This is important to consider when calculating the previous metrics such as popularity.

In [16]:
%%sql
SELECT 
	FORMAT((SUM(duration) / 1000) / 60 / 60, 2) AS DurationInHours
FROM surf_rock_sunshine srs;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
1 rows affected.


[('4.82',)]

The Surf Rock Sunshine playlist is approximately 4.82 hours in duration, which is actually quite short compared to many other playlists. The next step for TuneCore may be to encourage surf rock artists to pitch to this playlist in hopes of it generating them more streams.

### Primary Queries - Artist Rankings and Popularity

1. Because the fundamental question I am looking into is figuring out which playlists TuneCore artists should pitch their music to, I am using this query to determine the top ranking artists based on popularity in the Ultimate Indie playlist. By figuring out the top artists on this playlist, it provides insight for TuneCore that they can use for their online tutoring modules.

To figure out the top three artists with the highest average popularities alongside their ranking in the Ultimate Indie playlist, I had to use a CTE, Window, and GROUP BY function.

In [25]:
%%sql
WITH ArtistRank AS (
	SELECT
		artist,
		FORMAT(AVG(popularity), 0) AS AveragePopularity,
		DENSE_RANK() OVER(
			ORDER BY AVG(popularity) DESC)
		AS PopularityRank
	FROM ultimate_indie ui
	GROUP BY artist
)
SELECT *
FROM ArtistRank
WHERE PopularityRank <= 3;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
3 rows affected.


[('TV Girl', '84', 1), ("Her's", '80', 2), ('Clairo', '77', 3)]

Based on the results of the query, it can be determined that TV Girl, Her's and Clairo are the top three ranking on the Ultimate Indie playlist. The next best step that can be taken here would be to have TuneCore reach out to these artists to ask them what some of the steps they (or their managers/producers) took to pitch their songs to the Spotify and earn them a spot on the playlist.

Another great next step in terms of data analysis would be to look into these artists' performance on other indie related playlists.

# ----------------------------------------------------------------------------------------

2. In this follow up query, I am diving into the question I previously asked from the primary query. Here, I am using a COUNT on a CASE function that will count tracks from the three artists in two other playlists, Surf Rock Sunshine and Bedroom Pop. I am using the UNION function in place of the JOIN function because each table in my database is assigned to a separate Spotify playlist, and therefore the UNION function joins all tracks together under the same attributes.

In [26]:
%%sql
SELECT
	CASE WHEN COUNT(CASE WHEN artist = 'Her''s' THEN name END) = 1 THEN 'Surf Rock Sunshine' END AS Playlist,
	COUNT(CASE WHEN artist = 'TV Girl' THEN name END) AS TVGirlTracks,
	COUNT(CASE WHEN artist = 'Her''s' THEN name END) AS HersTracks,
	COUNT(CASE WHEN artist = 'Clairo' THEN name END) AS ClairoTracks
FROM surf_rock_sunshine srs
UNION ALL
SELECT
	CASE WHEN COUNT(CASE WHEN artist = 'Clairo' THEN name END) = 1 THEN 'Bedroom Pop' END AS Playlist,
	COUNT(CASE WHEN artist = 'TV Girl' THEN name END) AS TVGirlTracks,
	COUNT(CASE WHEN artist = 'Her''s' THEN name END) AS HersTracks,
	COUNT(CASE WHEN artist = 'Clairo' THEN name END) AS ClairoTracks
FROM bedroom_pop bp;

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
2 rows affected.


[('Surf Rock Sunshine', 0, 1, 0), ('Bedroom Pop', 0, 0, 1)]

Based on the results, Her's also has a track on Surf Rock Sunshine, while Clairo has a track on Bedroom Pop. This is important to consider as these playlists may not cultivate a similar audience as the ones that listen to the Ultimate Indie playlist. Some next steps that could be taken here is having TuneCore warn their artists that not all playlists relating to a certain genre cultivate the same audience.

# ----------------------------------------------------------------------------------------

3. This final query dives into answering the average duration of the songs written by these three artists. To do this, the average durations are grouped by the artists and converted into minutes rather than milliseconds. In addition, a subquery is used to ensure that only the three artists we are looking for are returned.

In [28]:
%%sql
SELECT
	artist,
	FORMAT((AVG(duration) / 1000) / 60 , 2) AS AvgArtistDuration,
	FORMAT(AVG(popularity), 0) AS AvgPopularity
FROM ultimate_indie ui
WHERE artist IN
    (
        SELECT
        	artist
        FROM ultimate_indie ui2 
        WHERE 
        	artist = 'Clairo'
        	OR artist = 'Her''s'
        	OR artist = 'TV Girl'
    )
GROUP BY artist
ORDER BY FORMAT((AVG(duration) / 1000) / 60 , 2);

 * mysql://admin:***@lmu-dev-01.c43apdfbq0ti.us-west-1.rds.amazonaws.com/sql_project
3 rows affected.


[('TV Girl', '3.57', '84'), ('Clairo', '3.82', '77'), ("Her's", '4.25', '80')]

It can be determined here that between the three artists, their songs have an average duration between three and a half minutes to four and a half minutes. It could be inferred then, that their shorter songs catch the attention of audiences on the playlist and are fully played through. Some next steps might be to encourage TuneCore artists to strongly consider the length of their songs before pitching them to Spotify. This can be a huge factor into whether their music will be accepted or listened to all the way through. 